core/src/main/java/org/apache/iceberg/RemoveSnapshots.java

```diff
@@ -77,8 +79,34 @@ public ExpireSnapshots expireSnapshotId(long expireSnapshotId) {
 
   @Override
   public ExpireSnapshots expireOlderThan(long timestampMillis) {
-    LOG.info("Expiring snapshots older than: {} ({})", new Date(timestampMillis), timestampMillis);
-    this.expireOlderThan = timestampMillis;
+    updateExpireOlderThan(timestampMillis, true);
+    return this;
+  }
+
+  private boolean updateExpireOlderThan(long timestampMillis, boolean fromExpireOlderThan) {
+    if (expireOlderThan == null ||
+        expireOlderThan > timestampMillis ||
+        this.isExpireOlderThanSet == fromExpireOlderThan) {
+      LOG.info("Expiring snapshots older than: {} ({})", new Date(timestampMillis), timestampMillis);
+      this.expireOlderThan = timestampMillis;
+      this.isExpireOlderThanSet = fromExpireOlderThan;
+      return true;
+    }
+    return false;
+  }
+
+  @Override
+  public ExpireSnapshots retainLast(int numSnapshots) {
```

Comment:
I think the current implementation is confusing. The nth snapshot is found by traversing parents of the current snapshot, then any newer snapshots are kept. If I have a recently staged snapshot and run `retainLast(3)`, then the number of retained snapshots will be 4. The description does not document this behavior.

This also requires coordinating `expireOlderThan` and `retainLast`, and that makes this more complicated. I think that these two should be independent. Behavior should be:
* All snapshots older than the `expireOlderThan` time are removed, _unless_ . . .
* A snapshot is one of the most recent `numSnapshots` from `retainLast`

I think this method should build a set of snapshot IDs that are to be retained that are the most recent, ignoring parentIds.

Reply:
@rdblue one question, what should be the behavior if I just call `retainLast(3)` without  `expireOlderThan` and meanwhile a new snapshot is added?

Reply:
I think that `retainLast` should designate what to keep. If `expireOlderThan` is not called, then no snapshots will be expired.

Also, if we choose to default both of these settings from table properties (e.g. age in days to expire), then the behavior would be to use the table settings as the default and then override with these calls. If neither the table expiration age setting nor `expireOlderThan` are used, then no snapshots would expire.

Reply:
@rdblue interesting behavior. 

Your earlier comment _"I would expect to be able to combine this (`retainLast(20)`) with a call to `expireOlderThan(T)`, where snapshots older than T are expired, but at least 20 snapshots are kept."_ . 

I think the behavior you mentioned in your earlier comment is more straightforward. We could discuss more on the behavior of `retainLast(n)` and how to find the N snapshots.

Another way to expire snapshots is `expireOlderThanSnapshot(snapshotId)`.

Reply:
I'm describing the same behavior in both comments. What sounds different?

Reply:
@rdblue looks like I got it wrong. I thought of keeping `expireOlderThan` and  `retainLast` independent of each other, and either one can expire snapshots independently.

Does it makes sense to add support for `expireOlderThanSnapshot(snapshotId)`?

Reply:
No, I don't think so. What is the use case for `expireOlderThanSnapshot`?

Reply:
Use case is opposite of `expireSnapshotId`.

If you are experimenting with different optimizations like `commit.manifest.target-size-bytes` , `RewriteManifests.clusterBy` will end up with lots of manifest and snapshots. After you get the configurations correct, you can expire older snapshots.

Reply:
You can always use `expireOlderThan(someSnapshot.timestampMillis())`

Reply:
yeah I agree